README
[ this is -*-Text-*- ]

Complete documentation is given in info file.

This is librep, a Lisp system for UNIX, needed by Sawfish window
manager. It contains a Lisp interpreter, byte-code compiler and
virtual machine. Applications may use the Lisp interpreter as an
extension language, or it may be used for stand-alone scripts.

The Lisp dialect was originally inspired by Emacs Lisp, but with the
worst features removed. It also borrows many ideas from Scheme.

It is known to compile on at least Solaris/sparc and Linux, x86 and
x86-64.

It is released under the terms of the GNU GPL, 
copyright John Harper <jsh@pixelslut.com>

Maintainer until Aug 2008: John Harper
Maintainer since Aug 2008: Christopher Bratusek

For more details see:

	http://sawfish.wikia.com/

To compile this you'll need GNU make, and GNU dbm
installed. Basically, just do:

	$ ./configure
	$ make
	$ make install

Notes
=====

64-bit systems
--------------

Basically support of 64-bit systems works, but if not, read this:

You may want to look at the `--with-value-type' configure option. This
is an implicitly signed integer type (i.e. `int', `long', etc) that is
wide enough to store an arbitrary pointer without losing any bits.

It should be detected automatically by the configure script, but if not
there are two most likely required settings:

 1. For a machine with 64-bit pointers and longs, but only 32-bit ints
    the following could be done:

	$ ./configure --with-value-type=long

 2. For a machine with both int and long 32-bits, but with 64-bit
    pointers and long long ints, then:

	$ ./configure --with-value-type="long long"

If this option is set incorrectly (i.e. to an integer type that is too
small) a run-time assertion will be triggered when the interpreter
initialises itself.

Also, if this option is set to anything but int, long, or long long,
then the src/rep_config.h file will need to be edited for the constant
suffix and printf conversion of the chosen type.


libgmp
------

rep optionally uses GNU MP for its bignum/rational implementation;
you can find it at any GNU mirror. GMP versions 2 and 3 are both known
to work (though version 3 is recommended)


shared libraries
----------------

rep includes plugins providing language bindings for several libraries.
Because these plugins are implemented as shared objects that are loaded
at runtime, the libraries they wrap must also be shared libraries on
most systems. This means that the installed libgdm must be shared, and
if compiling with readline support, so must libreadline


cut-down version
----------------

The configure script accepts the following options to build restricted
versions of librep. The resulting library is binary-compatible with the
normal version.

  --without-gmp

    Don't use GNU MP for bignums. Use `long long' as biggest integer
    type (if available, else just `long'). Also, there is no support
    for exact rationals, thus (/ 1 2) => 0.5 not 1/2

  --disable-continuations

    Don't include support for call/cc or multi-threading. This may be
    useful for machines with non-linear stacks (old crays?)


obscure configure options
-------------------------

  --enable-dballoc

    Trace all memory allocations. Not for general use

  --disable-dbsyms

    When printing C stack backtraces, don't try to output symbolic
    addresses

  --enable-gprof

    Compile with flags enabling profiling. Also needs --enable-static
    to be given. Use the `srep' target in the src directory to build a
    statically linked interpreter (since gprof doesn't profile shared
    libraries)

  --with-malloc-alignment=BYTES

    The minimum alignment of memory returned from malloc (). Defaults
    to the machine's word size. It's unlikely this will ever need to be
    specified explicitly

  --with-stack-direction=DIRECTION

    Direction of stack growth. -1 for downwards (grows from higher
    addresses to lower addresses), +1 for upwards. If not given, will
    try to infer this automatically (though it has been known to fail)

  --enable-full-name-terminator=CHARACTER

    If the GECOS fields in your password file contain extra information
    after the user's full name, this option allows the separator
    character to be given, letting rep's user-full-name function return
    the correct information.

    E.g. some systems have GECOS as a comma-separated list of values,
    the first of which is the full name. For this case:

	--enable-full-name-terminator=,



ChangeLog
2013-01-17  Christopher Roy Bratusek <nano@tuxfamily.org>
	* src/lispmach.h: fix compilation on ARM [Togan Muftouglu]

2013-01-12  Christopher Roy Bratusek <nano@tuxfamily.org>
	* debian/changelog
	* debian/control
	* debian/librep16.symbols: misc updates to debian packaging scripts

2012-09-08  Christopher Roy Bratusek <nano@tuxfamily.org>
	* lisp/rep/util/misc.jl: add program-exists-p [Mark Triggs]

	* configure.in: bump version to 0.92.3

	* man/news.texi: update

2012-09-04  Christopher Roy Bratusek <nano@tuxfamily.org>
	* configure.in: fix stack-direction for s390 [Kim B. Heino]

2012-09-01  Christopher Roy Bratusek <nano@tuxfamily.org>
	* configure.in: fix stack-direction for ppc [Kim B. Heino]

	* install-sh: updated to 2011-01-19.21

	* man/news.texi: update

2012-07-06  Christopher Roy Bratusek <nano@tuxfamily.org>
	* lisp/rep/xml/reader.jl: more entities support: gt, umlauts, esszett

	* COPYING
	* lisp/rep.jl
	* lisp/rep/data.jl
	* lisp/rep/data/objects.jl
	* lisp/rep/data/queues.jl
	* lisp/rep/data/records.jl
	* lisp/rep/data/ring.jl
	* lisp/rep/data/sort.jl
	* lisp/rep/data/string-util.jl
	* lisp/rep/data/symbol-table.jl
	* lisp/rep/data/trie.jl
	* lisp/rep/ffi/util.jl
	* lisp/rep/i18n/xgettext.jl
	* lisp/rep/io/file-handlers.jl
	* lisp/rep/io/file-handlers/remote.jl
	* lisp/rep/io/file-handlers/remote/ftp.jl
	* lisp/rep/io/file-handlers/remote/rcp.jl
	* lisp/rep/io/file-handlers/remote/rep.jl
	* lisp/rep/io/file-handlers/remote/utils.jl
	* lisp/rep/io/file-handlers/tar.jl
	* lisp/rep/io/file-handlers/tilde.jl
	* lisp/rep/io/files.jl
	* lisp/rep/io/streams.jl
	* lisp/rep/lang/compat-doc.jl
	* lisp/rep/lang/debugger.jl
	* lisp/rep/lang/define.jl
	* lisp/rep/lang/doc.jl
	* lisp/rep/lang/error-helper.jl
	* lisp/rep/lang/interpreter.jl
	* lisp/rep/lang/math.jl
	* lisp/rep/lang/profiler.jl
	* lisp/rep/mail/addr.jl
	* lisp/rep/module-system.jl
	* lisp/rep/net/domain-name.jl
	* lisp/rep/net/rpc.jl
	* lisp/rep/regexp.jl
	* lisp/rep/structures.jl
	* lisp/rep/system.jl
	* lisp/rep/system/environ.jl
	* lisp/rep/system/pwd-prompt.jl
	* lisp/rep/test/autoload.jl
	* lisp/rep/test/data.jl
	* lisp/rep/test/framework.jl
	* lisp/rep/threads/condition-variable.jl
	* lisp/rep/threads/message-port.jl
	* lisp/rep/threads/mutex.jl
	* lisp/rep/threads/proxy.jl
	* lisp/rep/threads/utils.jl
	* lisp/rep/user.jl
	* lisp/rep/util/autoloader.jl
	* lisp/rep/util/base64.jl
	* lisp/rep/util/date.jl
	* lisp/rep/util/gaol.jl
	* lisp/rep/util/ispell.jl
	* lisp/rep/util/memoize.jl
	* lisp/rep/util/misc.jl
	* lisp/rep/util/repl.jl
	* lisp/rep/util/time.jl
	* lisp/rep/vm/assembler.jl
	* lisp/rep/vm/bytecode-defs.jl
	* lisp/rep/vm/bytecodes.jl
	* lisp/rep/vm/compiler.jl
	* lisp/rep/vm/compiler/basic.jl
	* lisp/rep/vm/compiler/bindings.jl
	* lisp/rep/vm/compiler/inline.jl
	* lis/lispmach.h
	* src/rep.h
	* src/rep_gh.h
	* src/rep_lisp.h
	* src/rep_subrs.h
	* src/repint.h
	* src/repint_subrs.h
	* src/unix_defs.h: fix fsf address [Togan Muftuoglu]

2012-03-24  Christopher Roy Bratusek <nano@tuxfamily.org>
	* src/streams.c: read_line no longer aborts at 400 characters [Timo Korvola]

2012-01-14  Christopher Roy Bratusek <nano@tuxfamily.org>
	* configure.in: fixed build-failure with gcc >= 4.7.0 on x86_64 [Kim B. Heino]

2012-01-13  Christopher Roy Bratusek <nano@tuxfamily.org>
	* librep.spec: sync with Fedora [Kim B. Heino]

2011-08-20  Christopher Roy Bratusek <nano@tuxfamily.org>
	* src/Makefile.in
	* lisp/Makefile.in: Fix build on OS-X 10.6 by not exporting librep.sym
	via libtool [John Harper]

	* src/librep.sym: added some missing symbols [John Harper]

	* librep.pc.in: don't ignore datarootdir

	* src/ffi.c: better version of ALIGN if undefined [John Harper]

	* configure.in: check for ffi/ffi.h if ffi.h isn't found [John Harper]

2011-06-10  Christopher Roy Bratusek <nano@tuxfamily.org>
	* lisp/rep/util/misc.jl: imported beautify-symbol-name and
	remove-newlines and option-index from Sawfish.
	Improved version of position by Eli Barzilay

2011-06-02  Christopher Roy Bratusek <nano@tuxfamily.org>
	* lisp/rep/util/misc.jl: added file containing misc stuff imported
	from Sawfish. currently `position' and `string->symbol'

2011-05-05  Christopher Roy Brautsek <nano@tuxfamily.org>
	* **/Makefile.in: remove VPATH to appease makepp [Daniel Pfeiffer]

2011-04-14  Christopher Bratusek <nano@tuxfamily.org>
	* configure.in
	* Makefile.in: remove version from install-paths

	* debian/librep16.install: updated

2011-04-13  Christopher Bratusek <nano@tuxfamily.org>
	* Makedefs.in
	* Makefile.in
	* configure.in
	* src/misc.c
	* build-info
	* librep.pc.in: removed architecture from install-paths

	* Makefile.in: added debclean rule
	               improved distclean rule
		       improved gitclean rule

	* debian/**: merged packaing scripts from Ian Zimmermann

	* config.h.in: removed from GIT

	* config.guess
	* config.sub: added

2011-04-05  Christopher Bratusek <nano@tuxfamily.org>
	* src/unix_dl.c
	* src/lispcmds.c: load shared-objects rather than libtool-archvies
	 -- Kim B. Heino

2011-03-18  Christopher Bratusek <nano@tuxfamily.org>
	* man/rep.texi: fixed a typo


2011-02-27  Christopher Bratusek <zanghar@freenet.de>
	* debian/watch
	* debian/changelog.in
	* debian/control.in
	* debian/README.source: removed

	* configure.in: don't substitude gitdate
	                don't generate debian/changelog
			don't generate debian/control

	* debian/rules: major update

	* debian/rep-doc.docs
	* debian/control
	* debian/changelog: added

	* debian/rep.links
	* debian/librep-dbg.links
	* debian/librep-dev.links: updated

	* debian/patches/00list
	* debian/patches/series: renamed file

	* debian/source/format: changed to 3.0 (quilt)

2011-02-05  Teika Kazura <teika@lavabit.com>
	* lisp/rep/system.jl
	* man/lang.texi: Bugfix of "remove-hook-by-name"; exported the
	function.  This fixes the bug in commit
	c7da736dbdf0cb8b70d0289ebdda03036ec71add.  Also bugfix in
	lang.texi is included.

	* lisp/rep/lang/interpreter.jl

src/md5.c
/* md5.c - Functions to compute MD5 message digest of files or memory blocks
   according to the definition of MD5 in RFC 1321 from April 1992.
   Copyright (C) 1995, 1996 Free Software Foundation, Inc.
   NOTE: The canonical source of this file is maintained with the GNU C
   Library.  Bugs can be reported to bug-glibc@prep.ai.mit.edu.

   This program is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by the
   Free Software Foundation; either version 2, or (at your option) any
   later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.  */

/* Written by Ulrich Drepper <drepper@gnu.ai.mit.edu>, 1995.  */

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#include <sys/types.h>

#if STDC_HEADERS || defined _LIBC
# include <stdlib.h>
# include <string.h>
#else
# ifndef HAVE_MEMCPY
#  define memcpy(d, s, n) bcopy ((s), (d), (n))
# endif
#endif

#include "md5.h"

#ifdef _LIBC
# include <endian.h>
# if __BYTE_ORDER == __BIG_ENDIAN
#  define WORDS_BIGENDIAN 1
# endif
#endif

#ifdef WORDS_BIGENDIAN
# define SWAP(n)							\
    (((n) << 24) | (((n) & 0xff00) << 8) | (((n) >> 8) & 0xff00) | ((n) >> 24))
#else
# define SWAP(n) (n)
#endif


/* This array contains the bytes used to pad the buffer to the next
   64-byte boundary.  (RFC 1321, 3.1: Step 1)  */
static const unsigned char fillbuf[64] = { 0x80, 0 /* , 0, 0, ...  */ };


/* Initialize structure containing state of computation.
   (RFC 1321, 3.3: Step 3)  */
void
md5_init_ctx (ctx)
     struct md5_ctx *ctx;
{
  ctx->A = 0x67452301;
  ctx->B = 0xefcdab89;
  ctx->C = 0x98badcfe;
  ctx->D = 0x10325476;

  ctx->total[0] = ctx->total[1] = 0;
  ctx->buflen = 0;
}

/* Put result from CTX in first 16 bytes following RESBUF.  The result
   must be in little endian byte order.

   IMPORTANT: On some systems it is required that RESBUF is correctly
   aligned for a 32 bits value.  */
void *
md5_read_ctx (ctx, resbuf)
     const struct md5_ctx *ctx;
     void *resbuf;
{
  ((md5_uint32 *) resbuf)[0] = SWAP (ctx->A);
  ((md5_uint32 *) resbuf)[1] = SWAP (ctx->B);
  ((md5_uint32 *) resbuf)[2] = SWAP (ctx->C);
  ((md5_uint32 *) resbuf)[3] = SWAP (ctx->D);

  return resbuf;
}

/* Process the remaining bytes in the internal buffer and the usual
   prolog according to the standard and write the result to RESBUF.

   IMPORTANT: On some systems it is required that RESBUF is correctly
   aligned for a 32 bits value.  */
void *
md5_finish_ctx (ctx, resbuf)
     struct md5_ctx *ctx;
     void *resbuf;
{
  /* Take yet unprocessed bytes into account.  */
  md5_uint32 bytes = ctx->buflen;
  size_t pad;

  /* Now count remaining bytes.  */
  ctx->total[0] += bytes;
  if (ctx->total[0] < bytes)
    ++ctx->total[1];

  pad = bytes >= 56 ? 64 + 56 - bytes : 56 - bytes;
  memcpy (&ctx->buffer[bytes], fillbuf, pad);

  /* Put the 64-bit file length in *bits* at the end of the buffer.  */
  *(md5_uint32 *) &ctx->buffer[bytes + pad] = SWAP (ctx->total[0] << 3);
  *(md5_uint32 *) &ctx->buffer[bytes + pad + 4] = SWAP ((ctx->total[1] << 3) |
							(ctx->total[0] >> 29));

  /* Process last bytes.  */
  md5_process_block (ctx->buffer, bytes + pad + 8, ctx);

  return md5_read_ctx (ctx, resbuf);
}

/* Compute MD5 message digest for bytes read from STREAM.  The
   resulting message digest number will be written into the 16 bytes
   beginning at RESBLOCK.  */
int
md5_stream (stream, resblock)
     FILE *stream;
     void *resblock;
{
  /* Important: BLOCKSIZE must be a multiple of 64.  */
#define BLOCKSIZE 4096
  struct md5_ctx ctx;
  char buffer[BLOCKSIZE + 72];
  size_t sum;

  /* Initialize the computation context.  */
  md5_init_ctx (&ctx);

  /* Iterate over full file contents.  */
  while (1)
    {
      /* We read the file in blocks of BLOCKSIZE bytes.  One call of the
	 computation function processes the whole buffer so that with the
	 next round of the loop another block can be read.  */
      size_t n;
      sum = 0;

      /* Read block.  Take care for partial reads.  */
      do
	{
	  n = fread (buffer + sum, 1, BLOCKSIZE - sum, stream);

	  sum += n;
	}
      while (sum < BLOCKSIZE && n != 0);
      if (n == 0 && ferror (stream))
        return 1;

      /* If end of file is reached, end the loop.  */
      if (n == 0)
	break;

      /* Process buffer with BLOCKSIZE bytes.  Note that
			BLOCKSIZE % 64 == 0
       */
      md5_process_block (buffer, BLOCKSIZE, &ctx);
    }

  /* Add the last bytes if necessary.  */
  if (sum > 0)
    md5_process_bytes (buffer, sum, &ctx);

  /* Construct result in desired memory.  */
  md5_finish_ctx (&ctx, resblock);
  return 0;
}

/* Compute MD5 message digest for LEN bytes beginning at BUFFER.  The
   result is always in little endian byte order, so that a byte-wise
   output yields to the wanted ASCII representation of the message
   digest.  */
void *
md5_buffer (buffer, len, resblock)
     const char *buffer;
     size_t len;
     void *resblock;
{
  struct md5_ctx ctx;

  /* Initialize the computation context.  */
  md5_init_ctx (&ctx);

  /* Process whole buffer but last len % 64 bytes.  */
  md5_process_bytes (buffer, len, &ctx);

  /* Put result in desired memory area.  */
  return md5_finish_ctx (&ctx, resblock);
}


void
md5_process_bytes (buffer, len, ctx)
     const void *buffer;
     size_t len;
     struct md5_ctx *ctx;
{
  /* When we already have some bits in our internal buffer concatenate
     both inputs first.  */
  if (ctx->buflen != 0)
    {
      size_t left_over = ctx->buflen;
      size_t add = 128 - left_over > len ? len : 128 - left_over;

      memcpy (&ctx->buffer[left_over], buffer, add);
      ctx->buflen += add;

      if (left_over + add > 64)
	{
	  md5_process_block (ctx->buffer, (left_over + add) & ~63, ctx);
	  /* The regions in the following copy operation cannot overlap.  */
	  memcpy (ctx->buffer, &ctx->buffer[(left_over + add) & ~63],
		  (left_over + add) & 63);
	  ctx->buflen = (left_over + add) & 63;
	}

      buffer = (const char *) buffer + add;
      len -= add;
    }

  /* Process available complete blocks.  */
  if (len > 64)
    {
      md5_process_block (buffer, len & ~63, ctx);
      buffer = (const char *) buffer + (len & ~63);
      len &= 63;
    }

  /* Move remaining bytes in internal buffer.  */
  if (len > 0)
    {
      memcpy (ctx->buffer, buffer, len);
      ctx->buflen = len;
    }
}


/* These are the four functions used in the four steps of the MD5 algorithm
   and defined in the RFC 1321.  The first function is a little bit optimized
   (as found in Colin Plumbs public domain implementation).  */
/* #define FF(b, c, d) ((b & c) | (~b & d)) */
#define FF(b, c, d) (d ^ (b & (c ^ d)))
#define FG(b, c, d) FF (d, b, c)
#define FH(b, c, d) (b ^ c ^ d)
#define FI(b, c, d) (c ^ (b | ~d))

/* Process LEN bytes of BUFFER, accumulating context into CTX.
   It is assumed that LEN % 64 == 0.  */

void
md5_process_block (buffer, len, ctx)
     const void *buffer;
     size_t len;
     struct md5_ctx *ctx;
{
  md5_uint32 correct_words[16];
  const md5_uint32 *words = buffer;
  size_t nwords = len / sizeof (md5_uint32);
  const md5_uint32 *endp = words + nwords;
  md5_uint32 A = ctx->A;
  md5_uint32 B = ctx->B;
  md5_uint32 C = ctx->C;
  md5_uint32 D = ctx->D;

  /* First increment the byte count.  RFC 1321 specifies the possible
     length of the file up to 2^64 bits.  Here we only compute the
     number of bytes.  Do a double word increment.  */
  ctx->total[0] += len;
  if (ctx->total[0] < len)
    ++ctx->total[1];

  /* Process all bytes in the buffer with 64 bytes in each round of
     the loop.  */
  while (words < endp)
    {
      md5_uint32 *cwp = correct_words;
      md5_uint32 A_save = A;
      md5_uint32 B_save = B;
      md5_uint32 C_save = C;
      md5_uint32 D_save = D;

      /* First round: using the given function, the context and a constant
	 the next context is computed.  Because the algorithms processing
	 unit is a 32-bit word and it is determined to work on words in
	 little endian byte order we perhaps have to change the byte order
	 before the computation.  To reduce the work for the next steps
	 we store the swapped words in the array CORRECT_WORDS.  */

#define OP(a, b, c, d, s, T)						\
      do								\
        {								\
	  a += FF (b, c, d) + (*cwp++ = SWAP (*words)) + T;		\
	  ++words;							\
	  CYCLIC (a, s);						\
	  a += b;							\
        }								\
      while (0)

      /* It is unfortunate that C does not provide an operator for
	 cyclic rotation.  Hope the C compiler is smart enough.  */
#define CYCLIC(w, s) (w = (w << s) | (w >> (32 - s)))

      /* Before we start, one word to the strange constants.
	 They are defined in RFC 1321 as

	 T[i] = (int) (4294967296.0 * fabs (sin (i))), i=1..64
       */

      /* Round 1.  */
      OP (A, B, C, D,  7, 0xd76aa478);
      OP (D, A, B, C, 12, 0xe8c7b756);
      OP (C, D, A, B, 17, 0x242070db);
      OP (B, C, D, A, 22, 0xc1bdceee);
      OP (A, B, C, D,  7, 0xf57c0faf);
      OP (D, A, B, C, 12, 0x4787c62a);
      OP (C, D, A, B, 17, 0xa8304613);
      OP (B, C, D, A, 22, 0xfd469501);
      OP (A, B, C, D,  7, 0x698098d8);
      OP (D, A, B, C, 12, 0x8b44f7af);
      OP (C, D, A, B, 17, 0xffff5bb1);
      OP (B, C, D, A, 22, 0x895cd7be);
      OP (A, B, C, D,  7, 0x6b901122);
      OP (D, A, B, C, 12, 0xfd987193);
      OP (C, D, A, B, 17, 0xa679438e);
      OP (B, C, D, A, 22, 0x49b40821);

      /* For the second to fourth round we have the possibly swapped words
	 in CORRECT_WORDS.  Redefine the macro to take an additional first
	 argument specifying the function to use.  */
#undef OP
#define OP(f, a, b, c, d, k, s, T)					\
      do 								\
	{								\
	  a += f (b, c, d) + correct_words[k] + T;			\
	  CYCLIC (a, s);						\
	  a += b;							\
	}								\
      while (0)

      /* Round 2.  */
      OP (FG, A, B, C, D,  1,  5, 0xf61e2562);
      OP (FG, D, A, B, C,  6,  9, 0xc040b340);
      OP (FG, C, D, A, B, 11, 14, 0x265e5a51);
      OP (FG, B, C, D, A,  0, 20, 0xe9b6c7aa);
      OP (FG, A, B, C, D,  5,  5, 0xd62f105d);
      OP (FG, D, A, B, C, 10,  9, 0x02441453);
      OP (FG, C, D, A, B, 15, 14, 0xd8a1e681);
      OP (FG, B, C, D, A,  4, 20, 0xe7d3fbc8);
      OP (FG, A, B, C, D,  9,  5, 0x21e1cde6);
      OP (FG, D, A, B, C, 14,  9, 0xc33707d6);
      OP (FG, C, D, A, B,  3, 14, 0xf4d50d87);
      OP (FG, B, C, D, A,  8, 20, 0x455a14ed);
      OP (FG, A, B, C, D, 13,  5, 0xa9e3e905);
      OP (FG, D, A, B, C,  2,  9, 0xfcefa3f8);
      OP (FG, C, D, A, B,  7, 14, 0x676f02d9);
      OP (FG, B, C, D, A, 12, 20, 0x8d2a4c8a);

      /* Round 3.  */
      OP (FH, A, B, C, D,  5,  4, 0xfffa3942);
      OP (FH, D, A, B, C,  8, 11, 0x8771f681);
      OP (FH, C, D, A, B, 11, 16, 0x6d9d6122);
      OP (FH, B, C, D, A, 14, 23, 0xfde5380c);
      OP (FH, A, B, C, D,  1,  4, 0xa4beea44);
      OP (FH, D, A, B, C,  4, 11, 0x4bdecfa9);
      OP (FH, C, D, A, B,  7, 16, 0xf6bb4b60);
      OP (FH, B, C, D, A, 10, 23, 0xbebfbc70);
      OP (FH, A, B, C, D, 13,  4, 0x289b7ec6);
      OP (FH, D, A, B, C,  0, 11, 0xeaa127fa);
      OP (FH, C, D, A, B,  3, 16, 0xd4ef3085);
      OP (FH, B, C, D, A,  6, 23, 0x04881d05);
      OP (FH, A, B, C, D,  9,  4, 0xd9d4d039);
      OP (FH, D, A, B, C, 12, 11, 0xe6db99e5);
      OP (FH, C, D, A, B, 15, 16, 0x1fa27cf8);
      OP (FH, B, C, D, A,  2, 23, 0xc4ac5665);

      /* Round 4.  */
      OP (FI, A, B, C, D,  0,  6, 0xf4292244);
      OP (FI, D, A, B, C,  7, 10, 0x432aff97);
      OP (FI, C, D, A, B, 14, 15, 0xab9423a7);
      OP (FI, B, C, D, A,  5, 21, 0xfc93a039);
      OP (FI, A, B, C, D, 12,  6, 0x655b59c3);
      OP (FI, D, A, B, C,  3, 10, 0x8f0ccc92);
      OP (FI, C, D, A, B, 10, 15, 0xffeff47d);
      OP (FI, B, C, D, A,  1, 21, 0x85845dd1);
      OP (FI, A, B, C, D,  8,  6, 0x6fa87e4f);
      OP (FI, D, A, B, C, 15, 10, 0xfe2ce6e0);
      OP (FI, C, D, A, B,  6, 15, 0xa3014314);
      OP (FI, B, C, D, A, 13, 21, 0x4e0811a1);
      OP (FI, A, B, C, D,  4,  6, 0xf7537e82);
      OP (FI, D, A, B, C, 11, 10, 0xbd3af235);
      OP (FI, C, D, A, B,  2, 15, 0x2ad7d2bb);
      OP (FI, B, C, D, A,  9, 21, 0xeb86d391);

      /* Add the starting values of the context.  */
      A += A_save;
      B += B_save;
      C += C_save;
      D += D_save;
    }

  /* Put checksum in context given as argument.  */
  ctx->A = A;
  ctx->B = B;
  ctx->C = C;
  ctx->D = D;
}
